lib/FindTests.js

```diff
@@ -0,0 +1,269 @@
+) /*: Promise<Array<{ moduleName: string, possiblyTests: Array<string> }>> */ {
+  return Promise.all(
+    testFilePaths.map((filePath) => {
+      const matchingSourceDirs = project.testsSourceDirs.filter((dir) =>
+        filePath.startsWith(`${dir}${path.sep}`)
+      );
+
+      // Tests must be in tests/ or in source-directories – otherwise they won’t
+      // compile. Elm won’t be able to find imports.
+      switch (matchingSourceDirs.length) {
+        case 0:
+          return Promise.reject(
+            Error(
+              missingSourceDirectoryError(
+                filePath,
+                project.elmJson.type === 'package'
+              )
+            )
+          );
+
+        case 1:
+          // Keep going.
+          break;
+
+        default:
+          // This shouldn’t be possible for package projects.
+          return Promise.reject(
+            new Error(
+              multipleSourceDirectoriesError(
+                filePath,
+                matchingSourceDirs,
+                project.testsDir
+              )
+            )
+          );
+      }
+
+      // By finding the module name from the file path we can import it even if
+      // the file is full of errors. Elm will then report what’s wrong.
+      const moduleNameParts = path
+        .relative(matchingSourceDirs[0], filePath)
+        .replace(/\.elm$/, '')
+        .split(path.sep);
+      const moduleName = moduleNameParts.join('.');
+
+      if (!moduleNameParts.every(Parser.isUpperName)) {
+        return Promise.reject(
+          new Error(
+            badModuleNameError(filePath, matchingSourceDirs[0], moduleName)
+          )
+        );
+      }
+
+      return Parser.extractExposedPossiblyTests(
+        filePath,
+        // We’re reading files asynchronously in a loop here, so it makes sense
+        // to use graceful-fs to avoid “too many open files” errors.
+        gracefulFs.createReadStream
+      ).then((possiblyTests) => ({
+        moduleName,
+        possiblyTests,
+      }));
+    })
+  );
+}
+
+function missingSourceDirectoryError(filePath, isPackageProject) {
+  return `
+This file:
+
+${filePath}
+
+…matches no source directory! Imports won't work then.
+
+${
+  isPackageProject
+    ? 'Move it to tests/ or src/ in your project root.'
+    : 'Move it to tests/ in your project root, or make sure it is covered by "source-directories" in your elm.json.'
+}
+  `.trim();
+}
+
+function multipleSourceDirectoriesError(
+  filePath,
+  matchingSourceDirs,
+  testsDir
+) {
+  const note = matchingSourceDirs.includes(testsDir)
+    ? "Note: The tests/ folder counts as a source directory too (even if it isn't listed in your elm.json)!"
+    : '';
+
+  return `
+This file:
+
+${filePath}
+
+…matches more than one source directory:
+
+${matchingSourceDirs.join('\n')}
+
+Edit "source-directories" in your elm.json and try to make it so no source directory contains another source directory!
+
+${note}
+  `.trim();
+}
```

Comment:
From Runner.js. The only change is adding the note about the tests/ folder counting as a source directory too. That happens when the user has `"."` in their source directories, which elm-test does not support if you want to use the tests/ directory. See also: https://discourse.elm-lang.org/t/elm-test-error-with-directory-name-structure/6552